File: curso/modulo-05/aula-16/assinatura_digital_pem.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;


Console.WriteLine("========================== CARREGANDO CHAVE PRIVADA =========================");
ECDsa ec = ECDsa.Create(ECCurve.NamedCurves.nistP256);
var pem = File.ReadAllText("ecc-private-key.pem");
ec.ImportFromPem(pem);
var parametros = ec.ExportParameters(true);
var detalhesParametros = new
{
    Curve = "P-256",
    parametros.D,
    Q = new
    {
        parametros.Q.X,
        parametros.Q.Y

    }
};
Console.WriteLine($"Informacoes da chave:");
Console.WriteLine(JsonSerializer.Serialize(detalhesParametros, new JsonSerializerOptions() { WriteIndented = true }));

Console.WriteLine("========================== ASSINAR DIGITALMENTE =========================");
var assinatura = ec.SignData(Encoding.UTF8.GetBytes("desenvolvedor.io"), HashAlgorithmName.SHA256);
Console.WriteLine($"Assinatura: {Convert.ToHexString(assinatura)}");


Console.WriteLine("========================== CARREGAR CHAVE PUBLICA =========================");

ec = ECDsa.Create();
pem = File.ReadAllText("ecc-public-key.pem");
ec.ImportFromPem(pem);

Console.WriteLine("Utilizando objeto ECDsa apenas com chave publica");


Console.WriteLine("========================== VALIDAR ASSINATURA COM CHAVE PUBLICA =========================");
Console.WriteLine($"Assinatura esta valida: {ec.VerifyData(Encoding.UTF8.GetBytes("desenvolvedor.io"), assinatura, HashAlgorithmName.SHA256)}");

File: curso/modulo-05/aula-16/assinatura_digital_pem.py
from ecdsa import SigningKey, VerifyingKey

print("Carregando chave privada")
with open('ecc-private-key.pem', 'r') as pem:
    assinante = SigningKey.from_pem(pem.read())

mensagem = b'desenvolvedorio'

print('Assinante PEM:')
print(assinante.to_pem().decode())

assinatura = assinante.sign(mensagem)

print("Carregando chave publica")
with open('ecc-public-key.pem', 'r') as pem:
    verificador = VerifyingKey.from_pem(pem.read())

print('Verificador PEM:')
print(verificador.to_pem().decode())

print("Assinatura valida:", verificador.verify(assinatura, mensagem))
